plugins/manager/manager.go

```diff
@@ -0,0 +1,242 @@
+	pluginInstances     map[plugins.PluginID]PluginInstance
+
+	// plugin contains all the information needed to launch and dispense the
+	// Nomad Autoscaler plugins.
+	pluginsLock sync.RWMutex
+	plugins     map[plugins.PluginID]*pluginInfo
+}
+
+// pluginInfo contains all the required information to launch an Autoscaler
+// plugin. Not all details are required depending on whether the plugin is
+// internal or external; these are noted on the params.
+type pluginInfo struct {
+	baseInfo *plugins.PluginInfo
+	config   map[string]string
+
+	// args and exePath are required to execute the external plugin command.
+	args    []string
+	exePath string
+
+	// factory is only populated when the plugin is internal.
+	factory plugins.PluginFactory
+}
+
+// NewPluginManager sets up a new PluginManager for use.
+func NewPluginManager(log hclog.Logger, dir string, cfg map[string][]*config.Plugin) *PluginManager {
+	return &PluginManager{
+		cfg:             generateConfigMap(cfg),
+		logger:          log.Named("plugin-manager"),
+		pluginDir:       dir,
+		pluginInstances: make(map[plugins.PluginID]PluginInstance),
+		plugins:         make(map[plugins.PluginID]*pluginInfo),
+	}
+}
+
+// Load is responsible for registering and executing the plugins configured for
+// use by the Autoscaler agent.
+func (pm *PluginManager) Load() error {
+
+	for t, cfgs := range pm.cfg {
+		for _, cfg := range cfgs {
+
+			// Figure out if the plugin is internal or external, then perform
+			// the loading of the config into the manager store.
+			if pm.useInternal(cfg.Driver) {
+				pm.loadInternalPlugin(cfg, t)
+			} else {
+				pm.loadExternalPlugin(cfg, t)
+			}
+		}
+	}
+
+	return pm.dispensePlugins()
+}
+
+// KillPlugins calls Kill on all plugins currently dispensed.
+func (pm *PluginManager) KillPlugins() {
+	for id, v := range pm.pluginInstances {
+		pm.logger.Info("shutting down plugin", "plugin_name", id.Name)
+		v.Kill()
+	}
+}
+
+// Dispense returns a PluginInstance for use by safely obtaining the
+// PluginInstance from storage if we have it.
+func (pm *PluginManager) Dispense(name, pluginType string) (PluginInstance, error) {
+	pm.pluginInstancesLock.RLock()
+	defer pm.pluginInstancesLock.RUnlock()
+
+	// Attempt to pull our plugin instance from the store and pass this to the
+	// caller.
+	//
+	// TODO(jrasell) if we do not find the instance, we should probably try and
+	//  dispense the plugin. We should also check the plugin instance has not
+	//  exited.
+	inst, ok := pm.pluginInstances[plugins.PluginID{Name: name, PluginType: pluginType}]
+	if !ok {
+		return nil, fmt.Errorf("failed to dispense plugin: %s of type %s is not stored", name, pluginType)
+	}
+	return inst, nil
+}
+
+// dispensePlugins launches all configured plugins. It is responsible for
+// executing external binaries as well as setting the config on all plugins so
+// they are in a ready state. Any errors from this process will result in the
+// agent failing on startup.
```

Comment:
Yeah I agree. I went with this route for the PR as its very clear. If we want to just log these errors and continue the startup it is a single line change. It would be worth noting for the future to present this errors in a nicer way to the operator.

Reply:
I would say in this case it's OK since it's part of the agent startup flow, so it might be better to fail early because the fix here is most likely updating a config and restarting the agent, meaning all policies would be disrupted anyway. If we fail early small errors would be easier to detect.

A similar example is for nginx config files, if you have even one upstream wrong the server won't start (which I have a love-hate relationship with, but it's usually my fault for not checking with `-t`).

We could reevaluate this once we have signal config reload, better error visibility, more detailed health check etc.